### PR TITLE
Feat(trainer): Add elastic dataloader with changeable batch size.

### PR DESCRIPTION
**Pull request by Antlera**
### What changes were proposed in this pull request?

Add elastic dataloader with variable batch size.

### Why are the changes needed?

Efficient training needs changeable batch size in runtime.

### Does this PR introduce any user-facing change?

People can use `ElasticDataloader` to replace `Dataloader` in standard torch utils to enable training with dynamic batch size.

### How was this patch tested?

Tested by UT include in this pr.

### CODE REVIEW

File: dlrover/trainer/torch/elastic_dataloader.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+from torch.utils.data import BatchSampler, DataLoader
+
+
+class ElasticDataLoader(DataLoader):
```
Comment (workingloong) [inline dlrover/trainer/torch/elastic_dataloader.py:19]: Add a docstring to explain the `ElasticDataLoader` like https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L123-L211

File: dlrover/trainer/torch/elastic_dataloader.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self, *args, config_file=None, **kwargs):
+        super(ElasticDataLoader, self).__init__(*args, **kwargs)
+        self.current_batch_size = self.batch_size
+        self.config_file = config_file
+
+        if self.config_file:
+            self.load_config()
+
+    def load_config(self):
```
Comment (workingloong) [inline dlrover/trainer/torch/elastic_dataloader.py:28]:
```
def load_config(self, config_file=None):
    if not config_file:
        return
    with open(config_file, "r") as f:
       ...
```
The refactor can reduce indent levels and we and test the `local_config` without changing the class.

File: dlrover/trainer/torch/elastic_dataloader.py (hunk 1 above)
Comment (merlintang) [reply dlrover/trainer/torch/elastic_dataloader.py:19]: +1

**Review by merlintang — APPROVED** [review]

**Review by workingloong — APPROVED** [review]
LGTM

**Pull request merged**